zai_sapi/testing/catch2.hpp

```diff
@@ -0,0 +1,290 @@
+#ifndef HAVE_ZAI_SAPI_TEST
+#define HAVE_ZAI_SAPI_TEST
+/**
+* ZAI SAPI Test Harness in Pre Processor:
+*
+*  ZAI SAPI is tested with catch2 and this header provides a convenience
+*  and consistency API over catch2
+*
+*  Test cases will all have meaningful naming (screen vs code), and sensible
+*  form in test files.
+*
+*  For convenience:
+*    ZAI_SAPI_TEST_CASE[_WITH_TAGS]_BARE(char *suite, char *description, [char *tags, ] block_t code)
+*
+*    Will produce a case with tagging consistent with other tests but not decorate the code
+*
+*  The decorator API provided is essentially:
+*    ZAI_SAPI_TEST_[BAILING_]CASE[_WITH_TAGS][_WITH_STUB][_WITH_PROLOGUE]
+*       (char *suite, char *description, [char *tags,] [char *stub,] [block_t prologue,] block_t code)
+*
+*    Delete [] as appropriate
+*
+* This completes the master class in naming things, you are ready ...
+*
+* Notes:
+*  Tests not declared to bail that bail will fail
+*
+*  Tests declared to bail that do not bail will fail
+*
+*  Stubs (as a pre-requisite of test code) are required to execute without error
+*    Errors in stubs will cause the test to fail
+*/
+extern "C" {
+#include "../zai_sapi.h"
+}
+
+#include <catch2/catch.hpp>
+// clang-format off
+#define ZAI_SAPI_TEST_TAGS_NONE     ""
+#define ZAI_SAPI_TEST_STUB_NONE     NULL
+#define ZAI_SAPI_TEST_PROLOGUE_NONE {}
+
+/* {{{ private void ZAI_SAPI_TEST_CASE_TAG(char *suite, char *description, char *tags) */
+#define ZAI_SAPI_TEST_CASE_TAG(\
+        __ZAI_SAPI_TEST_CASE_SUITE,                                       \
+        __ZAI_SAPI_TEST_CASE_DESCRIPTION,                                 \
+        __ZAI_SAPI_TEST_CASE_TAGS)                                        \
+    __ZAI_SAPI_TEST_CASE_SUITE " [" __ZAI_SAPI_TEST_CASE_DESCRIPTION "]", \
+    "[" __ZAI_SAPI_TEST_CASE_SUITE "]" __ZAI_SAPI_TEST_CASE_TAGS /* }}} */
+
+/* {{{ public void ZAI_SAPI_TEST_CASE_BARE(char *suite, char *description, block_t code) */
+#define ZAI_SAPI_TEST_CASE_BARE(suite, description, ...)                     \
+    TEST_CASE(ZAI_SAPI_TEST_CASE_TAG(                                        \
+        suite, description, ZAI_SAPI_TEST_TAGS_NONE)) {                      \
+        { __VA_ARGS__ }                                                      \
+    } /* }}} */
+
+/* {{{ public void ZAI_SAPI_TEST_CASE_WITH_TAGS_BARE(char *suite, char *description, char *tags, block_t code) */
+#define ZAI_SAPI_TEST_CASE_WITH_TAGS_BARE(suite, description, tags, ...)     \
+    TEST_CASE(ZAI_SAPI_TEST_CASE_TAG(suite, description, tags)) {            \
+        { __VA_ARGS__ }                                                      \
+    } /* }}} */
+
+/* {{{ private bailout handling */
+#define __ZAI_SAPI_TEST_CASE_WITHOUT_BAILOUT_BEGIN()  \
+    bool zai_sapi_test_case_without_bailout = true;   \
```

Comment:
I just realized that all the `bool`s around the try/catches need to be marked as `volatile` so that they don't get clobbered by the long jump.